File: android/leokit/src/main/kotlin/io/github/leoho0722/leokit/components/LKBanner.kt
package io.github.leoho0722.leokit.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import io.github.leoho0722.leokit.LKShapes
import io.github.leoho0722.leokit.LKTheme
import io.github.leoho0722.leokit.tokens.LKSize
import io.github.leoho0722.leokit.tokens.LKSpacing

/** 橫幅提示要傳達的意思，決定它的底色與文字顏色。 */
public enum class LKBannerTone {
    /** 單純說明一件事，沒有好壞。 */
    Neutral,

    /** 補充說明或使用提示。 */
    Info,

    /** 事情做完了、設定生效了。 */
    Success,

    /** 要留意的狀況，還可以繼續用。 */
    Warning,

    /** 出錯了，使用者得先處理才能繼續。 */
    Danger,
}

/**
 * 留在版面上的提示，用來說明一個一直存在的狀況。
 *
 * 會自己消失的提示不能承載一定要被讀到的訊息；那是 Toast 的工作。
 * 一個畫面最多一個橫幅；同時有多個狀況時只顯示最嚴重的那一個。
 * 使用者自己解決不了的問題不要給 [onDismiss]，給了就等於允許他忽略它。
 *
 * @param message 說明這個狀況的內容
 * @param modifier 由呼叫端套在整個橫幅上的修飾子
 * @param title 一行標題，只有訊息很長時才需要
 * @param tone 要傳達的意思，決定底色與文字顏色
 * @param icon 開頭的圖示；不放圖示時傳 null
 * @param onDismiss 按下關閉時要做的事；傳 null 表示這個提示不能被關掉
 * @param dismissContentDescription 關閉鈕唸給螢幕閱讀器聽的說明
 * @param actions 橫幅下方的動作按鈕；不需要時傳 null
 */
@Composable
public fun LKBanner(
    message: String,
    modifier: Modifier = Modifier,
    title: String? = null,
    tone: LKBannerTone = LKBannerTone.Neutral,
    icon: ImageVector? = null,
    onDismiss: (() -> Unit)? = null,
    dismissContentDescription: String = "關閉提示",
    actions: (@Composable RowScope.() -> Unit)? = null,
) {
    val colors = LKTheme.colors
    val typography = LKTheme.typography
    val container: Color = when (tone) {
        LKBannerTone.Neutral -> colors.bgSubtle
        LKBannerTone.Info -> colors.brandSubtle
        LKBannerTone.Success -> colors.successSubtle
        LKBannerTone.Warning -> colors.warningSubtle
        LKBannerTone.Danger -> colors.dangerSubtle
    }
    val content: Color = when (tone) {
        LKBannerTone.Neutral -> colors.textPrimary
        LKBannerTone.Info -> colors.brandInk
        LKBannerTone.Success -> colors.successInk
        LKBannerTone.Warning -> colors.warningInk
        LKBannerTone.Danger -> colors.dangerInk
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(container, LKShapes.md)
            .padding(horizontal = LKSpacing.spacing16, vertical = LKSpacing.spacing12)
            .semantics { liveRegion = if (tone == LKBannerTone.Danger) LiveRegionMode.Assertive else LiveRegionMode.Polite },
        horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing12),
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = content,
                modifier = Modifier.size(LKSize.iconMd),
            )
        }
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(LKSpacing.spacing4),
        ) {
            if (title != null) {
                Text(text = title, style = typography.headline, color = content)
            }
            Text(text = message, style = typography.callout, color = content)
            if (actions != null) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing8),
                    verticalAlignment = Alignment.CenterVertically,
                    content = actions,
                )
            }
        }
        if (onDismiss != null) {
            IconButton(onClick = onDismiss, modifier = Modifier.size(LKSize.iconLg)) {
                Icon(
                    imageVector = LKIcons.Close,
                    contentDescription = dismissContentDescription,
                    tint = content,
                    modifier = Modifier.size(LKSize.iconSm),
                )
            }
        }
    }
}

File: apple/Sources/LeoKit/Components/LKBanner.swift
//
//  LKBanner.swift
//  LeoKit
//
//  Created by Leo Ho on 2026/09/19.
//

import SwiftUI

/// 留在版面上的提示，用來說明一個一直存在的狀況
///
/// - Note: 會自己消失的提示不能承載一定要被讀到的訊息，那是 Toast 的工作
/// - Note: 一個畫面最多一個橫幅，同時有多個狀況時只顯示最嚴重的那一個
/// - Note: 使用者自己解決不了的問題不要給 `onDismiss`，給了就等於允許他忽略它
public struct LKBanner<Actions: View>: View {

    // MARK: - Properties

    /// 橫幅下方的動作按鈕，不需要時傳 `EmptyView`
    private let actions: Actions

    /// 說明這個狀況的內容
    private let message: String

    /// 按下關閉時要做的事；傳 nil 表示這個提示不能被關掉
    private let onDismiss: (() -> Void)?

    /// 開頭的圖示名稱，取自 SF Symbols；不放圖示時是 nil
    private let systemImage: String?

    /// 一行標題，只有訊息很長時才需要
    private let title: String?

    /// 要傳達的意思，決定底色與文字顏色
    private let tone: LKBannerTone

    // MARK: - Init

    /// 建立一個留在版面上的提示
    ///
    /// - Parameters:
    ///   - message: 說明這個狀況的內容
    ///   - title: 一行標題
    ///   - tone: 要傳達的意思
    ///   - systemImage: 開頭的圖示名稱，取自 SF Symbols
    ///   - onDismiss: 按下關閉時要做的事
    ///   - actions: 橫幅下方的動作按鈕
    public init(
        _ message: String,
        title: String? = nil,
        tone: LKBannerTone = .neutral,
        systemImage: String? = nil,
        onDismiss: (() -> Void)? = nil,
        @ViewBuilder actions: () -> Actions
    ) {
        self.message = message
        self.title = title
        self.tone = tone
        self.systemImage = systemImage
        self.onDismiss = onDismiss
        self.actions = actions()
    }

    // MARK: - Body

    /// 橫幅的骨架：左邊圖示、中間文字與動作、右邊關閉鈕
    public var body: some View {
        HStack(alignment: .top, spacing: LKSpacing.spacing12) {
            icon

            content

            dismissButton
        }
        .padding(.horizontal, LKSpacing.spacing16)
        .padding(.vertical, LKSpacing.spacing12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .foregroundStyle(foreground)
        .background(background, in: LKShape.md)
        .accessibilityElement(children: .contain)
    }
}

// MARK: - Private Views

private extension LKBanner {

    /// 開頭的圖示，沒給名稱時不佔位置
    @ViewBuilder
    var icon: some View {
        if let systemImage {
            Image(systemName: systemImage)
                .font(.system(size: LKSize.iconMd))
        }
    }

    /// 中間的文字與動作按鈕，由上往下是標題、訊息、按鈕
    var content: some View {
        VStack(alignment: .leading, spacing: LKSpacing.spacing4) {
            titleText

            messageText

            actionButtons
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// 一行標題，沒給時不佔位置
    @ViewBuilder
    var titleText: some View {
        if let title {
            Text(title)
                .font(LKFont.headline)
        }
    }

    /// 說明這個狀況的內容
    var messageText: some View {
        Text(message)
            .font(LKFont.callout)
    }

    /// 橫幅下方的動作按鈕，一律用低調樣式，顏色跟著整個橫幅
    var actionButtons: some View {
        actions
            .padding(.top, LKSpacing.spacing4)
            .buttonStyle(.lk(.plain, size: .small))
    }

    /// 右上角的關閉鈕，不能關掉的提示不會顯示它
    @ViewBuilder
    var dismissButton: some View {
        if let onDismiss {
            Button {
                onDismiss()
            } label: {
                Image(systemName: "xmark")
                    .font(.system(size: LKSize.iconSm, weight: .semibold))
            }
            .buttonStyle(.plain)
            .accessibilityLabel("關閉提示")
        }
    }
}

// MARK: - Computed Properties

private extension LKBanner {

    /// 橫幅的底色
    var background: Color {
        switch tone {
        case .neutral:
            LKColor.bgSubtle
        case .info:
            LKColor.brandSubtle
        case .success:
            LKColor.successSubtle
        case .warning:
            LKColor.warningSubtle
        case .danger:
            LKColor.dangerSubtle
        }
    }

    /// 文字、圖示與裡面按鈕共用的顏色
    var foreground: Color {
        switch tone {
        case .neutral:
            LKColor.textPrimary
        case .info:
            LKColor.brandInk
        case .success:
            LKColor.successInk
        case .warning:
            LKColor.warningInk
        case .danger:
            LKColor.dangerInk
        }
    }
}

// MARK: - Convenience

extension LKBanner where Actions == EmptyView {

    /// 建立一個沒有動作按鈕的提示
    ///
    /// - Parameters:
    ///   - message: 說明這個狀況的內容
    ///   - title: 一行標題
    ///   - tone: 要傳達的意思
    ///   - systemImage: 開頭的圖示名稱，取自 SF Symbols
    ///   - onDismiss: 按下關閉時要做的事
    public init(
        _ message: String,
        title: String? = nil,
        tone: LKBannerTone = .neutral,
        systemImage: String? = nil,
        onDismiss: (() -> Void)? = nil
    ) {
        self.init(
            message,
            title: title,
            tone: tone,
            systemImage: systemImage,
            onDismiss: onDismiss
        ) {
            EmptyView()
        }
    }
}

// MARK: - Preview

#Preview("有標題與動作") {
    VStack(spacing: LKSpacing.spacing16) {
        LKBanner(
            "這個月的免費額度只剩三次。",
            title: "額度即將用完",
            tone: .warning,
            systemImage: "exclamationmark.triangle"
        ) {
            Button("查看方案") { }
        }

        LKBanner("備份已在昨天完成。", tone: .success, systemImage: "checkmark.circle")
    }
    .padding(LKSpacing.spacing16)
}
